src/coreclr/src/jit/codegencommon.cpp

```diff
@@ -4445,23 +4466,54 @@ void CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg, bool* pXtraRegClobbere
-                // SIMD floating point register.
-                argRegCount          = 2;
-                int       nextArgNum = argNum + 1;
-                regNumber nextRegNum = genMapRegArgNumToRegNum(nextArgNum, regArgTab[nextArgNum].getRegType(compiler));
-                noway_assert(regArgTab[nextArgNum].varNum == varNum);
-                noway_assert(genIsValidIntReg(nextRegNum));
-                noway_assert(genIsValidFloatReg(destRegNum));
-                GetEmitter()->emitIns_R_R_I(INS_mov, EA_8BYTE, destRegNum, nextRegNum, 1);
-            }
+                // This includes both fixed-size SIMD types that are independently promoted, as well
+                // as other HFA structs.
+                argRegCount = varDsc->lvHfaSlots();
+                if (argNum < (argMax - argRegCount + 1))
+                {
+                    if (compiler->lvaGetPromotionType(varDsc) == Compiler::PROMOTION_TYPE_INDEPENDENT)
+                    {
+                        // For an HFA type that is passed in multiple registers and promoted, we copy each field to its
+                        // destination register.
+                        for (int i = 0; i < argRegCount; i++)
+                        {
+                            int        nextArgNum  = argNum + i;
+                            LclVarDsc* fieldVarDsc = compiler->lvaGetDesc(varDsc->lvFieldLclStart + i);
+                            regNumber  nextRegNum =
+                                genMapRegArgNumToRegNum(nextArgNum, regArgTab[nextArgNum].getRegType(compiler));
+                            destRegNum = fieldVarDsc->GetRegNum();
+                            noway_assert(regArgTab[nextArgNum].varNum == varNum);
+                            noway_assert(genIsValidFloatReg(nextRegNum));
+                            noway_assert(genIsValidFloatReg(destRegNum));
+                            GetEmitter()->emitIns_R_R(INS_mov, EA_8BYTE, destRegNum, nextRegNum);
+                        }
+                    }
+#if defined(TARGET_ARM64) && defined(FEATURE_SIMD)
+                    else
+                    {
+                        // For a SIMD type that is passed in multiple registers but enregistered as a vector,
+                        // the code above copies the first argument register into the lower 4 or 8 bytes
+                        // of the target register. Here we must handle the subsequent fields by
+                        // inserting them into the upper bytes of the target SIMD floating point register.
+                        argRegCount = varDsc->lvHfaSlots();
+                        for (int i = 1; i < argRegCount; i++)
+                        {
+                            int       nextArgNum = argNum + i;
+                            regNumber nextRegNum =
+                                genMapRegArgNumToRegNum(nextArgNum, regArgTab[nextArgNum].getRegType(compiler));
```

Comment:
Nit: extract `regArgTab[nextArgNum]` to avoid the line break and use it in the next line as well.